File: apple/poc-contentfilter/Shared/YouTubeNormalize.swift
import Foundation

/// Swift mirror of `shared/youtube/youtube-normalize.ts`. The TypeScript file is
/// the authoritative spec; this must produce identical canonical objects so the
/// iOS content filter keys policy the same way as the backend and other adapters.
///
/// Approving one video must never widen to channel/recommendations/Shorts/search
/// — that guarantee is in the policy evaluation order, not here. This only maps a
/// URL to the canonical object it denotes.

public enum YouTubeKind: String {
    case video, channel, playlist, watchWithPlaylist, search, shorts, other
}

public struct YouTubeObject {
    public var kind: YouTubeKind
    public var videoId: String?
    public var channelId: String?
    public var channelHandle: String?
    public var playlistId: String?
    public var isYouTube: Bool
    public var host: String?
}

public enum YouTube {

    /// Hosts serving the YouTube application surface.
    static let appHosts: Set<String> = [
        "youtube.com", "m.youtube.com", "music.youtube.com",
        "gaming.youtube.com", "youtubekids.com", "youtube-nocookie.com",
    ]
    static let shortHosts: Set<String> = ["youtu.be"]

    /// Hosts required for an APPROVED video to actually play. Never block these
    /// while a video is approved. `*.googlevideo.com` is the opaque media CDN;
    /// allow the host while any video is approved and gate access at the watch
    /// page (which is per-video). See ARCHITECTURE.md §YouTube.
    public static let playbackSupportHosts: [String] = [
        "www.youtube.com", "youtubei.googleapis.com", "i.ytimg.com",
        "s.ytimg.com", "yt3.ggpht.com", "*.googlevideo.com",
        "jnn-pa.googleapis.com", "fonts.gstatic.com",
    ]

    static func isVideoId(_ s: String?) -> Bool {
        guard let s else { return false }
        return s.range(of: "^[A-Za-z0-9_-]{11}$", options: .regularExpression) != nil
    }
    static func isChannelId(_ s: String) -> Bool {
        s.range(of: "^UC[A-Za-z0-9_-]{22}$", options: .regularExpression) != nil
    }
    static func isPlaylistId(_ s: String) -> Bool {
        s.range(of: "^(?:PL|UU|LL|FL|RD|OL|EL)[A-Za-z0-9_-]{10,}$", options: .regularExpression) != nil
    }
    static func stripWww(_ host: String) -> String {
        var h = host.lowercased()
        if h.hasPrefix("www.") { h.removeFirst(4) }
        return h
    }

    public static func normalize(_ raw: String) -> YouTubeObject {
        guard let u = URLComponents(string: raw), let hostRaw = u.host else {
            return YouTubeObject(kind: .other, videoId: nil, channelId: nil,
                                 channelHandle: nil, playlistId: nil, isYouTube: false, host: nil)
        }
        let host = stripWww(hostRaw)
        func q(_ name: String) -> String? { u.queryItems?.first(where: { $0.name == name })?.value }
        let notYt = YouTubeObject(kind: .other, videoId: nil, channelId: nil,
                                  channelHandle: nil, playlistId: nil, isYouTube: false, host: host)

        // youtu.be/<id>
        if shortHosts.contains(host) {
            let seg = u.path.split(separator: "/").map(String.init)
            if let id = seg.first, isVideoId(id) {
                let list = q("list")
                return YouTubeObject(kind: list != nil ? .watchWithPlaylist : .video,
                                     videoId: id, channelId: nil, channelHandle: nil,
                                     playlistId: (list.map(isPlaylistId) == true) ? list : nil,
                                     isYouTube: true, host: host)
            }
            var o = notYt; o.isYouTube = true; return o
        }

        guard appHosts.contains(host) else { return notYt }

        let path = u.path.hasSuffix("/") ? String(u.path.dropLast()) : u.path
        let seg = path.split(separator: "/").map(String.init)
        let p0 = seg.first?.lowercased()

        switch p0 {
        case "watch":
            if let id = q("v"), isVideoId(id) {
                let list = q("list")
                return YouTubeObject(kind: list != nil ? .watchWithPlaylist : .video,
                                     videoId: id, channelId: nil, channelHandle: nil,
                                     playlistId: (list.map(isPlaylistId) == true) ? list : nil,
                                     isYouTube: true, host: host)
            }
        case "shorts" where isVideoId(seg[safe: 1]):
            return YouTubeObject(kind: .shorts, videoId: seg[1], channelId: nil, channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        case "embed", "v":
            if isVideoId(seg[safe: 1]) {
                return YouTubeObject(kind: .video, videoId: seg[1], channelId: nil, channelHandle: nil,
                                     playlistId: nil, isYouTube: true, host: host)
            }
            if p0 == "embed", seg[safe: 1]?.lowercased() == "videoseries", let list = q("list"), isPlaylistId(list) {
                return YouTubeObject(kind: .playlist, videoId: nil, channelId: nil, channelHandle: nil,
                                     playlistId: list, isYouTube: true, host: host)
            }
        case "live" where isVideoId(seg[safe: 1]):
            return YouTubeObject(kind: .video, videoId: seg[1], channelId: nil, channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        case "playlist":
            if let list = q("list"), isPlaylistId(list) {
                return YouTubeObject(kind: .playlist, videoId: nil, channelId: nil, channelHandle: nil,
                                     playlistId: list, isYouTube: true, host: host)
            }
        case "channel" where seg[safe: 1].map(isChannelId) == true:
            return YouTubeObject(kind: .channel, videoId: nil, channelId: seg[1], channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        case let p? where p.hasPrefix("@"):
            return YouTubeObject(kind: .channel, videoId: nil, channelId: nil, channelHandle: seg[0],
                                 playlistId: nil, isYouTube: true, host: host)
        case "user", "c":
            if let name = seg[safe: 1] {
                return YouTubeObject(kind: .channel, videoId: nil, channelId: nil,
                                     channelHandle: "\(p0!)/\(name)", playlistId: nil, isYouTube: true, host: host)
            }
        case "results":
            return YouTubeObject(kind: .search, videoId: nil, channelId: nil, channelHandle: nil,
                                 playlistId: nil, isYouTube: true, host: host)
        default:
            break
        }
        var o = notYt; o.isYouTube = true; return o
    }

    /// Stable policy key, e.g. "YOUTUBE_VIDEO:dQw4w9WgXcQ".
    public static func policyKey(_ o: YouTubeObject) -> String? {
        if let v = o.videoId { return "YOUTUBE_VIDEO:\(v)" }
        if let pl = o.playlistId, o.kind == .playlist { return "YOUTUBE_PLAYLIST:\(pl)" }
        if let c = o.channelId { return "YOUTUBE_CHANNEL:\(c)" }
        if let h = o.channelHandle { return "YOUTUBE_CHANNEL_HANDLE:\(h)" }
        return nil
    }
}

private extension Array {
    subscript(safe i: Int) -> Element? { indices.contains(i) ? self[i] : nil }
}

File: apple/AjarSafari/Extension/youtube-normalize.js
/**
 * youtube-normalize.js — JS port of `shared/youtube/youtube-normalize.ts`.
 *
 * The TypeScript file is the AUTHORITATIVE SPEC (ADR-007, ADR-008). This port
 * must produce identical canonical objects so the Safari extension keys policy
 * exactly the same way as the backend, the iOS Swift port, and the Windows
 * extension. Keep it in LOCKSTEP with the TS — treat any divergence as a bug.
 *
 * Approving one video must never widen to the channel / recommendations /
 * Shorts / search — that guarantee lives in the evaluation order (background.js
 * / policy-model.ts), not here. This only maps a URL to the object it denotes.
 *
 * Dependency-free ES module. Imported by background.js (a MV3 service worker
 * declared as type:"module" in manifest.json).
 */

/** Hosts that serve the YouTube application surface (watch pages, shorts, etc.). */
const YT_APP_HOSTS = new Set([
  "youtube.com",
  "m.youtube.com",
  "music.youtube.com",
  "gaming.youtube.com",
  "youtubekids.com",
  "youtube-nocookie.com", // privacy-enhanced embeds
]);

/** Short-link host: the entire path is the video id. */
const YT_SHORT_HOSTS = new Set(["youtu.be"]);

/**
 * Resource hosts REQUIRED for an approved video to actually play. These are NOT
 * policy targets and must NEVER be blocked while a video is approved. In
 * particular `*.googlevideo.com` is the opaque, per-session media CDN — allow
 * the whole host while ANY video is currently approved on the device and rely
 * on the per-video watch-page gate. See ARCHITECTURE.md §6 and docs/MACOS_SAFARI_POC.md B7.
 *
 * Kept identical to YOUTUBE_PLAYBACK_SUPPORT_HOSTS in the TypeScript.
 */
export const YOUTUBE_PLAYBACK_SUPPORT_HOSTS = [
  "www.youtube.com", // InnerTube /youtubei/v1/player API, base JS
  "youtubei.googleapis.com", // InnerTube API (mobile/native players)
  "i.ytimg.com", // thumbnails / storyboards
  "s.ytimg.com", // player javascript/css
  "yt3.ggpht.com", // avatars/thumbs
  "*.googlevideo.com", // MEDIA STREAMS (opaque; see note above)
  "jnn-pa.googleapis.com", // BotGuard/attestation used by the player
  "fonts.gstatic.com", // player glyphs
];

const VIDEO_ID_RE = /^[A-Za-z0-9_-]{11}$/;
const CHANNEL_ID_RE = /^UC[A-Za-z0-9_-]{22}$/;
const PLAYLIST_ID_RE = /^(?:PL|UU|LL|FL|RD|OL|EL)[A-Za-z0-9_-]{10,}$/;

function stripWww(host) {
  return host.replace(/\.$/, "").replace(/^www\./i, "").toLowerCase();
}

function isValidVideoId(id) {
  return !!id && VIDEO_ID_RE.test(id);
}

/**
 * Reduce a URL to its canonical YouTube object. Returns `{ isYouTube:false }`
 * for non-YouTube URLs. Never throws. Mirrors normalizeYouTube() in the TS.
 */
export function normalizeYouTube(rawUrl) {
  let u;
  try {
    u = new URL(rawUrl);
  } catch {
    return { kind: "other", isYouTube: false };
  }

  const host = stripWww(u.hostname);
  const notYt = { kind: "other", isYouTube: false, host };

  // youtu.be/<id>
  if (YT_SHORT_HOSTS.has(host)) {
    const seg = u.pathname.split("/").filter(Boolean);
    const id = seg[0];
    if (isValidVideoId(id)) {
      const listId = u.searchParams.get("list") ?? undefined;
      return {
        kind: listId ? "watch_with_playlist" : "video",
        videoId: id,
        playlistId: listId && PLAYLIST_ID_RE.test(listId) ? listId : undefined,
        isYouTube: true,
        host,
      };
    }
    return { ...notYt, isYouTube: true };
  }

  if (!YT_APP_HOSTS.has(host)) return notYt;

  const path = u.pathname.replace(/\/+$/, ""); // trim trailing slashes
  const seg = path.split("/").filter(Boolean);
  const p0 = seg[0]?.toLowerCase();

  // /watch?v=<id>[&list=<pl>]
  if (p0 === "watch") {
    const id = u.searchParams.get("v");
    const listId = u.searchParams.get("list") ?? undefined;
    if (isValidVideoId(id)) {
      return {
        kind: listId ? "watch_with_playlist" : "video",
        videoId: id,
        playlistId: listId && PLAYLIST_ID_RE.test(listId) ? listId : undefined,
        isYouTube: true,
        host,
      };
    }
    return { ...notYt, isYouTube: true };
  }

  // /shorts/<id>
  if (p0 === "shorts" && isValidVideoId(seg[1])) {
    return { kind: "shorts", videoId: seg[1], isYouTube: true, host };
  }

  // /embed/<id>  and /v/<id>  (nocookie host lands here too)
  if ((p0 === "embed" || p0 === "v") && isValidVideoId(seg[1])) {
    return { kind: "video", videoId: seg[1], isYouTube: true, host };
  }
  // /embed/videoseries?list=<pl>
  if (p0 === "embed" && seg[1]?.toLowerCase() === "videoseries") {
    const listId = u.searchParams.get("list") ?? undefined;
    if (listId && PLAYLIST_ID_RE.test(listId))
      return { kind: "playlist", playlistId: listId, isYouTube: true, host };
  }

  // /live/<id>  (premieres / livestream permalinks)
  if (p0 === "live" && isValidVideoId(seg[1])) {
    return { kind: "video", videoId: seg[1], isYouTube: true, host };
  }

  // /playlist?list=<pl>
  if (p0 === "playlist") {
    const listId = u.searchParams.get("list") ?? undefined;
    if (listId && PLAYLIST_ID_RE.test(listId))
      return { kind: "playlist", playlistId: listId, isYouTube: true, host };
    return { ...notYt, isYouTube: true };
  }

  // /channel/<UC…>
  if (p0 === "channel" && seg[1] && CHANNEL_ID_RE.test(seg[1])) {
    return { kind: "channel", channelId: seg[1], isYouTube: true, host };
  }

  // /@handle  (modern channel handle)
  if (p0 && p0.startsWith("@")) {
    return { kind: "channel", channelHandle: seg[0], isYouTube: true, host };
  }

  // /user/<name>, /c/<name>  (legacy channel forms — resolve to channelId
  // server-side via the Data API; carried as a handle-like token meanwhile)
  if ((p0 === "user" || p0 === "c") && seg[1]) {
    return { kind: "channel", channelHandle: `${p0}/${seg[1]}`, isYouTube: true, host };
  }

  // /results?search_query=…
  if (p0 === "results") {
    return { kind: "search", isYouTube: true, host };
  }

  // Recognized YouTube surface we can't reduce to a specific object
  return { kind: "other", isYouTube: true, host };
}

/**
 * Stable policy key for a canonical object, e.g. "YOUTUBE_VIDEO:dQw4w9WgXcQ".
 * Mirrors youTubePolicyKey() in the TS. Adapters and the backend use this as
 * the lookup key against policy rules.
 */
export function youTubePolicyKey(o) {
  if (o.videoId) return `YOUTUBE_VIDEO:${o.videoId}`;
  if (o.playlistId && o.kind === "playlist") return `YOUTUBE_PLAYLIST:${o.playlistId}`;
  if (o.channelId) return `YOUTUBE_CHANNEL:${o.channelId}`;
  if (o.channelHandle) return `YOUTUBE_CHANNEL_HANDLE:${o.channelHandle}`;
  return null;
}

/**
 * True if `host` is one of the playback-support hosts (supports a leading
 * "*." wildcard, matching the CDN entry `*.googlevideo.com`). Used by
 * background.js to keep an approved video's media/streaming reachable (B7).
 */
export function isPlaybackSupportHost(host) {
  if (!host) return false;
  const h = host.replace(/\.$/, "").replace(/^www\./i, "").toLowerCase();
  return YOUTUBE_PLAYBACK_SUPPORT_HOSTS.some((entry) => {
    if (entry.startsWith("*.")) {
      const suffix = entry.slice(1); // ".googlevideo.com"
      return h === entry.slice(2) || h.endsWith(suffix);
    }
    return h === entry.replace(/\.$/, "").replace(/^www\./i, "");
  });
}
